From 9e6389011605f3acfd940292d1be49708b785c71 Mon Sep 17 00:00:00 2001
From: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
Date: Wed, 11 Sep 2013 12:08:40 +0200
Subject: [PATCH 310/509] dma: xilinx: vdma: Split
 xilinx_vdma_free_chan_resources() function

https://github.com/analogdevicesinc/linux.git xcomm_zynq
commit dc0dfe7fd06a25c863f150bc7f464faef54a7df6

The xilinx_vdma_free_chan_resources() function frees both the
descriptors and the descriptors pool. Split the descriptors freeing code
into a separate function to increase code reuse.

Signed-off-by: Laurent Pinchart <laurent.pinchart@ideasonboard.com>
Acked-by: Srikanth Thokala <sthokal@xilinx.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 drivers/dma/xilinx/xilinx_axivdma.c |   21 ++++++++++-----------
 1 files changed, 10 insertions(+), 11 deletions(-)

diff --git a/drivers/dma/xilinx/xilinx_axivdma.c b/drivers/dma/xilinx/xilinx_axivdma.c
index 1f99921..5fbcceee 100644
--- a/drivers/dma/xilinx/xilinx_axivdma.c
+++ b/drivers/dma/xilinx/xilinx_axivdma.c
@@ -281,17 +281,23 @@ static void xilinx_vdma_free_desc_list_reverse(struct xilinx_vdma_chan *chan,
 	}
 }
 
-static void xilinx_vdma_free_chan_resources(struct dma_chan *dchan)
+static void xilinx_vdma_free_descriptors(struct xilinx_vdma_chan *chan)
 {
-	struct xilinx_vdma_chan *chan = to_xilinx_chan(dchan);
 	unsigned long flags;
 
-	dev_dbg(chan->dev, "Free all channel resources.\n");
 	spin_lock_irqsave(&chan->lock, flags);
 	xilinx_vdma_free_desc_list(chan, &chan->active_list);
 	xilinx_vdma_free_desc_list(chan, &chan->pending_list);
 	spin_unlock_irqrestore(&chan->lock, flags);
+}
+
+static void xilinx_vdma_free_chan_resources(struct dma_chan *dchan)
+{
+	struct xilinx_vdma_chan *chan = to_xilinx_chan(dchan);
+
+	dev_dbg(chan->dev, "Free all channel resources.\n");
 
+	xilinx_vdma_free_descriptors(chan);
 	dma_pool_destroy(chan->desc_pool);
 	chan->desc_pool = NULL;
 }
@@ -881,18 +887,11 @@ fail:
 
 static void xilinx_vdma_terminate_all(struct xilinx_vdma_chan *chan)
 {
-	unsigned long flags;
-
 	/* Halt the DMA engine */
 	xilinx_vdma_halt(chan);
 
-	spin_lock_irqsave(&chan->lock, flags);
-
 	/* Remove and free all of the descriptors in the lists */
-	xilinx_vdma_free_desc_list(chan, &chan->pending_list);
-	xilinx_vdma_free_desc_list(chan, &chan->active_list);
-
-	spin_unlock_irqrestore(&chan->lock, flags);
+	xilinx_vdma_free_descriptors(chan);
 }
 
 static int xilinx_vdma_slave_config(struct xilinx_vdma_chan *chan,
-- 
1.7.5.4

